src/platform/PICO/bus_quadspi_pico.c

```diff
@@ -0,0 +1,371 @@
+#include "pico/bootrom.h"
+#include "hardware/regs/qmi.h"
+#include "hardware/structs/qmi.h"
+#include "hardware/xip_cache.h"
+#include "drivers/time.h"
+
+// Provide platform QUADSPI hardware table for RP2350 (single logical device)
+const quadSpiHardware_t quadSpiHardware[QUADSPIDEV_COUNT] = {
+    {
+        .device = QUADSPIDEV_1,
+        .reg = (QUADSPI_TypeDef *)0x1, // sentinel; common layer uses it only for identity
+        .clkPins = { { 0 } },
+        .bk1IO0Pins = { { 0 } },
+        .bk1IO1Pins = { { 0 } },
+        .bk1IO2Pins = { { 0 } },
+        .bk1IO3Pins = { { 0 } },
+        .bk1CSPins  = { { 0 } },
+        .bk2IO0Pins = { { 0 } },
+        .bk2IO1Pins = { { 0 } },
+        .bk2IO2Pins = { { 0 } },
+        .bk2IO3Pins = { { 0 } },
+        .bk2CSPins  = { { 0 } },
+    },
+};
+
+// Pin configure is a no-op on RP2350, QMI pins are fixed and controlled by ROM/QMI
+void quadSpiPinConfigure(const quadSpiConfig_t *pConfig)
+{
+    UNUSED(pConfig);
+    quadSpiDevice[QUADSPIDEV_1].dev = quadSpiHardware[0].reg;
+}
+
+static inline void pico_qspi_enter_cmd_mode(void)
+{
+    // Ensure XIP cache is clean before exiting XIP
+    xip_cache_clean_all();
+    rom_connect_internal_flash_fn connect_fn = (rom_connect_internal_flash_fn)rom_func_lookup_inline(ROM_FUNC_CONNECT_INTERNAL_FLASH);
+    rom_flash_exit_xip_fn exit_fn = (rom_flash_exit_xip_fn)rom_func_lookup_inline(ROM_FUNC_FLASH_EXIT_XIP);
+    connect_fn();
+    exit_fn();
+}
+
+static inline void pico_qspi_exit_cmd_mode(void)
+{
+    flash_flush_cache();
+}
+
+// QMI direct helpers for CS1 transactions
+static inline void qmi_direct_enable(void) { hw_set_bits(&qmi_hw->direct_csr, QMI_DIRECT_CSR_EN_BITS); }
+static inline void qmi_direct_disable(void) { hw_clear_bits(&qmi_hw->direct_csr, QMI_DIRECT_CSR_EN_BITS); }
+static inline void qmi_cs1_assert(bool asserted) {
+    if (asserted) hw_set_bits(&qmi_hw->direct_csr, QMI_DIRECT_CSR_ASSERT_CS1N_BITS);
+    else hw_clear_bits(&qmi_hw->direct_csr, QMI_DIRECT_CSR_ASSERT_CS1N_BITS);
+}
+
+static inline void qmi_timeout_cleanup(void)
+{
+    qmi_cs1_assert(false);
+    qmi_direct_disable();
+    pico_qspi_exit_cmd_mode();
+}
+
+// Default timeout for QSPI direct IO operations
+#define QSPI_TIMEOUT_MS 10
+
+// Number of dummy clock cycles per consumed byte in direct-mode loops.
+// Justification: QSPI datasheets specify dummy cycles in bits. Our direct-mode
+// helper consumes dummy time by shifting out whole bytes (8 bits at a time).
+// We therefore convert from bits to bytes using 8 bits/byte. If a controller
+// revision were to support sub-byte dummy handling, this could be revisited.
+#define QSPI_DUMMY_BITS_PER_BYTE 8
+
+static inline bool qmi_direct_io(const uint8_t *tx, uint8_t *rx, size_t count, timeMs_t timeout)
+{
+    size_t tx_remaining = count;
+    size_t rx_remaining = count;
+    // Prevent hangs: break out after timeout in ms
+    const timeMs_t start_ms = millis();
+    while (tx_remaining || rx_remaining) {
```

Comment:
(technically you can test only rx_remaining. And it may be even slightly easier to read  - loop until all replies are received)

Reply:
True, reception of the last byte indicates the completion of the IO operation.